File: wrappers/javascript/shared/src/store/OpenSession.ts
import type { StoreHandle } from '../crypto'

import { ariesAskar } from '../ariesAskar'
import { AriesAskarError } from '../error'

import { Session } from './Session'

export class OpenSession {
  private store: StoreHandle
  private profile?: string
  private isTxn: boolean
  // TODO: implement session
  private session?: Session

  public constructor({ store, isTxn, profile }: { store: StoreHandle; profile?: string; isTxn: boolean }) {
    this.store = store
    this.isTxn = isTxn
    this.profile = profile
  }

  public async open() {
    if (!this.store) throw AriesAskarError.customError({ message: 'Cannot start session from closed store' })
    if (this.session) throw AriesAskarError.customError({ message: 'Session already opened' })
    const sessionHandle = await ariesAskar.sessionStart({
      profile: this.profile,
      asTransaction: this.isTxn,
      storeHandle: this.store,
    })
    return new Session({ isTxn: this.isTxn, handle: sessionHandle })
  }
}

File: wrappers/kotlin/src/commonMain/kotlin/Store/OpenSession.kt
@file:OptIn(UnsafeNumber::class)

package askar.Store

import askar.Askar
import askar.crypto.SessionHandle
import askar.crypto.StoreHandle
import kotlinx.cinterop.UnsafeNumber
import kotlinx.coroutines.async
import kotlinx.coroutines.runBlocking

@OptIn(UnsafeNumber::class)
class OpenSession(private val store: StoreHandle, private val profile: String? = null, private val isTxn: Boolean) {
    //TODO: Implement session not done in javascript yet
    private var session: SessionHandle? = null

    suspend fun open(): Session {
        if (this.session != null) throw Error("Session already opened")
        val sessionHandle = Askar.session.sessionStart(store.handle, profile, isTxn)
        return Session( sessionHandle, this.isTxn )
    }
}